File: DuDoong-Api/src/main/java/band/gosrock/api/ticketItem/dto/response/TicketItemResponse.java
package band.gosrock.api.ticketItem.dto.response;


import band.gosrock.domain.common.vo.AccountInfoVo;
import band.gosrock.domain.common.vo.Money;
import band.gosrock.domain.domains.ticket_item.domain.TicketItem;
import band.gosrock.domain.domains.ticket_item.domain.TicketPayType;
import band.gosrock.domain.domains.ticket_item.domain.TicketType;
import io.swagger.v3.oas.annotations.media.Schema;
import lombok.Builder;
import lombok.Getter;

@Getter
@Builder
public class TicketItemResponse {
    @Schema(description = "티켓상품 id")
    private final Long ticketItemId;

    @Schema(description = "티켓 지불 타입")
    private final TicketPayType payType;

    @Schema(description = "이름")
    private final String ticketName;

    @Schema(description = "설명")
    private final String description;

    @Schema(description = "가격")
    private final Money price;

    @Schema(description = "티켓 승인 타입")
    private final TicketType approveType;

    @Schema(description = "1인당 구매 제한 매수")
    private final Long purchaseLimit;

    @Schema(description = "공급량")
    private final Long supplyCount;

    @Schema(description = "재고")
    private final Long quantity;

    @Schema(description = "재고공개 여부")
    private final Boolean isQuantityPublic;

    @Schema(description = "계좌 정보")
    private final AccountInfoVo accountInfo;

    public static TicketItemResponse from(TicketItem ticketItem, Boolean isAdmin) {

        return TicketItemResponse.builder()
                .ticketItemId(ticketItem.getId())
                .payType(ticketItem.getPayType())
                .ticketName(ticketItem.getName())
                .description(ticketItem.getDescription())
                .price(ticketItem.getPrice())
                .approveType(ticketItem.getType())
                .purchaseLimit(ticketItem.getPurchaseLimit())
                .supplyCount(ticketItem.getSupplyCount())
                .quantity(
                        isAdmin || ticketItem.getIsQuantityPublic()
                                ? ticketItem.getQuantity()
                                : null)
                .isQuantityPublic(ticketItem.getIsQuantityPublic())
                .accountInfo(ticketItem.toAccountInfoVo())
                .build();
    }
}

File: DuDoong-Api/src/main/kotlin/band/gosrock/api/ticketItem/dto/response/TicketItemResponse.kt
package band.gosrock.api.ticketItem.dto.response

import band.gosrock.domain.common.vo.AccountInfoVo
import band.gosrock.domain.common.vo.Money
import band.gosrock.domain.domains.ticket_item.domain.TicketItem
import band.gosrock.domain.domains.ticket_item.domain.TicketPayType
import band.gosrock.domain.domains.ticket_item.domain.TicketType
import io.swagger.v3.oas.annotations.media.Schema

data class TicketItemResponse(
    @field:Schema(description = "티켓상품 id")
    val ticketItemId: Long?,

    @field:Schema(description = "티켓 지불 타입")
    val payType: TicketPayType?,

    @field:Schema(description = "이름")
    val ticketName: String?,

    @field:Schema(description = "설명")
    val description: String?,

    @field:Schema(description = "가격")
    val price: Money?,

    @field:Schema(description = "티켓 승인 타입")
    val approveType: TicketType?,

    @field:Schema(description = "1인당 구매 제한 매수")
    val purchaseLimit: Long?,

    @field:Schema(description = "공급량")
    val supplyCount: Long?,

    @field:Schema(description = "재고")
    val quantity: Long?,

    @field:Schema(description = "재고공개 여부")
    val isQuantityPublic: Boolean?,

    @field:Schema(description = "계좌 정보")
    val accountInfo: AccountInfoVo?,

    @field:Schema(description = "재고가 감소한 티켓인지 리턴")
    val isSold: Boolean?,

    @field:Schema(description = "재고가 남아있는지 리턴")
    val isQuantityLeft: Boolean?,
) {
    companion object {
        @JvmStatic
        fun from(ticketItem: TicketItem, isAdmin: Boolean): TicketItemResponse = TicketItemResponse(
            ticketItemId = ticketItem.id,
            payType = ticketItem.payType,
            ticketName = ticketItem.name,
            description = ticketItem.description,
            price = ticketItem.price,
            approveType = ticketItem.type,
            purchaseLimit = ticketItem.purchaseLimit,
            supplyCount = ticketItem.supplyCount,
            quantity = if (isAdmin || ticketItem.isQuantityPublic == true) ticketItem.quantity else null,
            isQuantityPublic = ticketItem.isQuantityPublic,
            accountInfo = ticketItem.accountInfo,
            isSold = ticketItem.isSold(),
            isQuantityLeft = ticketItem.isQuantityLeft(),
        )
    }
}
